lgn-worker/src/main.rs

```diff
@@ -114,12 +129,195 @@ fn main() -> anyhow::Result<()> {
         .with_http_listener(([0, 0, 0, 0], config.prometheus.port))
         .install()?;
 
-    run(&config)
+    if let Some(grpc_url) = &config.avs.gateway_grpc_url {
+        run_with_grpc(&config, grpc_url).await?;
+    } else {
+        tokio::task::block_in_place(move || -> Result<()> {
+            run(&config)?;
+            Ok(())
+        })?;
+    }
+
+    Ok(())
 }
 
-fn run(config: &Config) -> Result<()> {
-    info!("Version: {}", env!("CARGO_PKG_VERSION"));
-    let lagrange_wallet = match (
+async fn maybe_download_checksum(config: &Config) -> Result<()> {
+    if config.public_params.skip_checksum {
+        return Ok(());
+    }
+
+    // Fetch checksum file
+    // The checksum file can be generated in two ways.
+    // 1- Run the worker, and it will download and spit out the checksum on disk
+    // 2- Manually download the params then install with the checksums bin crate and run checksums -c -r zkmr_params -a BLAKE3
+    let checksum_url = &config.public_params.checksum_url;
+    let expected_checksums_file = &config.public_params.checksum_expected_local_path;
+
+    let response = reqwest::get(checksum_url)
+        .await
+        .context("Failed to fetch checksum file")?
+        .text()
+        .await
+        .context("Failed to read response text")?;
+
+    let mut file = tokio::fs::File::create(expected_checksums_file)
+        .await
+        .context("Failed to create local checksum file")?;
+
+    file.write_all(response.as_bytes())
+        .await
+        .context("Failed to write checksum file")?;
+
+    drop(file);
+
+    Ok(())
+}
+
+async fn run_with_grpc(config: &Config, grpc_url: &str) -> Result<()> {
+    let uri = grpc_url.parse::<tonic::transport::Uri>()?;
+    let (mut outbound, outbound_rx) = tokio::sync::mpsc::channel(1024);
+
+    info!("Verifying the checksums");
+
+    maybe_download_checksum(config).await?;
+
+    let mut provers_manager =
+        tokio::task::block_in_place(move || -> Result<ProversManager<TaskType, ReplyType>> {
+            let mut provers_manager = ProversManager::<TaskType, ReplyType>::new();
+            register_v1_provers(config, &mut provers_manager)
+                .context("while registering provers")?;
+            Ok(provers_manager)
+        })?;
+
+    if !config.public_params.skip_checksum {
+        verify_directory_checksums(
+            &config.public_params.dir,
+            &config.public_params.checksum_expected_local_path,
+        )
+        .context("Failed to verify checksums")?;
+    }
```

Comment:
More general comment, but I think we should purely disable not checking the params, I don't see what would be the advantages of doing so. Wdyt @hackaugusto @kaisbaccour?

Reply:
Fair point, let's track this in a separate issue?